Add compiler switches for blaze versions

moxcodes opened the pull request:
- ensures compatibility with blaze 3.3 and 3.4, which change several typedefs, breaking
  features of PointerVector and DataVector.

## Proposed changes

### Types of changes:

- [ ] Bugfix

### Component:

- [ ] Code

### Code review checklist

- [ ] The PR passes all checks, including unit tests, `clang-tidy` and `IWYU`. For
  instructions on how to perform the CI checks locally refer to the [Dev guide
  on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run `make doc`
  to generate the documentation locally into `BUILD_DIR/docs/html`. Then open
  `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

nilsdeppe commented on src/DataStructures/DataVector.hpp:
```diff
@@ -58,8 +58,12 @@ using std::abs;  // NOLINT
 // IWYU pragma: no_include <blaze/math/traits/SubTrait.h>
 // IWYU pragma: no_include <blaze/system/TransposeFlag.h>
-// IWYU pragma: no_include <blaze/math/traits/BinaryMapTrait.h>
+#if ((BLAZE_MAJOR_VERSION == 3) && (BLAZE_MINOR_VERSION <= 3))
```
Maybe the more compact thing to do would be to add type aliases `blaze_enable_if` etc., use those and `ifdef` there?

moxcodes committed:
Add compiler switches for blaze versions
 - ensures compatibility with blaze 3.3 and 3.4, which change several typedefs,
  breaking features of PointerVector and DataVector.

moxcodes commented:
updated with a more compact set of switches. This way should also be easier to update in future if Blaze decides to change other things.

nilsdeppe reviewed (approved):
Much more maintainable :) Thanks for doing this!

kidder commented:
@wthrowe please look at

nilsdeppe commented:
Restart GCC debug build because of the random segfaults
